### PR TITLE
Fix/limitless dialect check + refactor synchronous router fetch

### PR DESCRIPTION
**Pull request by aaronchung-bitquill**
### Summary

Add checks for limitless dialect  + refactor synchronous router fetch #1148

### Description

- added dialect checks in mulitple places. 
- new code path incase dialect is not right, and trying to resolve it
- move limitless router endpoint fetching to a helper class `LimitlessQueryHelper`. 
- move synchronous fetch of limitless routers into service class
- move router cache out of monitor and into service
- add `put` method to SlidingExpirationCache
- added a check in DialectManager specifically for PG Limitless to resolve to Aurora PG dialect

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 1]
```diff
@@ -140,27 +157,29 @@ private Connection connectInternal(
...
+    }
+
+    if (limitlessRouters.contains(hostSpec)) {
       LOGGER.finest(Messages.get("LimitlessConnectionPlugin.connectWithHost", new Object[] {hostSpec.getHost()}));
       return connectFunc.call();
     }
 
     RoundRobinHostSelector.setRoundRobinHostWeightPairsProperty(props, limitlessRouters);
-    final HostSpec selectedHostSpec;
+    HostSpec selectedHostSpec;
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:173]: HostSpec selectedHostSpec =hostSpec ;

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 2]
```diff
@@ -140,27 +157,29 @@ private Connection connectInternal(
...
       selectedHostSpec = this.pluginService.getHostSpecByStrategy(limitlessRouters,
           HostRole.WRITER, RoundRobinHostSelector.STRATEGY_ROUND_ROBIN);
       LOGGER.fine(Messages.get(
           "LimitlessConnectionPlugin.selectedHost",
           new Object[] {selectedHostSpec.getHost()}));
     } catch (UnsupportedOperationException e) {
-      LOGGER.severe(Messages.get("LimitlessConnectionPlugin.incorrectConfiguration"));
-      throw e;
+      selectedHostSpec = hostSpec;
+      LOGGER.warning(Messages.get("LimitlessConnectionPlugin.errorSelectingRouter", new Object[] {e.getMessage()}));
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:182]: Such exception means that requested strategy isn't loaded. I think it's a showstopper.

Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:182]: You also need to catch SQLException. Some strategies raises SQLException when they can't get a host.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 3]
```diff
@@ -140,27 +157,29 @@ private Connection connectInternal(
       LOGGER.finest(Messages.get("LimitlessConnectionPlugin.limitlessRouterCacheEmpty"));
       final boolean waitForRouterInfo = WAIT_F0R_ROUTER_INFO.getBoolean(props);
       if (waitForRouterInfo) {
-        limitlessRouters = synchronouslyGetLimitlessRoutersWithRetry(props);
+        limitlessRouters = synchronouslyGetLimitlessRoutersWithRetry(connectFunc.call(), hostSpec.getPort(), props);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:160]: What if connectFunc.call() fails to connect?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 4]
```diff
@@ -140,27 +157,29 @@ private Connection connectInternal(
...
       } else {
         LOGGER.finest(Messages.get("LimitlessConnectionPlugin.usingProvidedConnectUrl"));
         return connectFunc.call();
       }
-    } else if (limitlessRouters.contains(hostSpec)) {
+    }
+
+    if (limitlessRouters.contains(hostSpec)) {
       LOGGER.finest(Messages.get("LimitlessConnectionPlugin.connectWithHost", new Object[] {hostSpec.getHost()}));
       return connectFunc.call();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:169]: You might have another connection already open at line 160. Can we return it?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 5]
```diff
@@ -171,29 +190,62 @@ private Connection connectInternal(
           new Object[] {selectedHostSpec.getHost()}));
       selectedHostSpec.setAvailability(HostAvailability.NOT_AVAILABLE);
       // Retry connect prioritising healthiest router for best chance of connection over load-balancing with round-robin
-      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, e);
+      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, connectFunc.call(), hostSpec.getPort(), e);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:193]: Another connection might have already been opened at line 160. Can we use it?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 6]
```diff
@@ -171,29 +190,62 @@ private Connection connectInternal(
...
+          .startMonitoring(hostSpec, properties, INTERVAL_MILLIS.getInteger(properties));
+    }
+
+    List<HostSpec> limitlessRouters = this.limitlessRouterService.getLimitlessRouters(
+        this.pluginService.getHostListProvider().getClusterId(), props);
+    if (limitlessRouters.isEmpty()) {
+      LOGGER.finest(Messages.get("LimitlessConnectionPlugin.limitlessRouterCacheEmpty"));
+      final boolean waitForRouterInfo = WAIT_F0R_ROUTER_INFO.getBoolean(props);
+      if (waitForRouterInfo) {
+        synchronouslyGetLimitlessRoutersWithRetry(connectFunc.call(), hostSpec.getPort(), props);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:223]: What if connectFunc.call() fails to connect?

Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:223]: Can we use connection from the line 204?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessQueryHelper.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    } catch (final SQLSyntaxErrorException e) {
+      throw new SQLException(Messages.get("LimitlessRouterMonitor.invalidQuery"), e);
+    } finally {
+      if (networkTimeout == 0 && !conn.isClosed()) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, networkTimeout);
+      }
+    }
+  }
+
+  private List<HostSpec> mapResultSetToHostSpecList(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessQueryHelper.java:78]: Can we make it protected to allow class extensibility?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessQueryHelper.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    long weight = Math.round(10 - cpu * 10);
+
+    if (weight < 1 || weight > 10) {
+      weight = 1; // default to 1
+      LOGGER.warning(() -> Messages.get(
+          "LimitlessRouterMonitor.invalidRouterLoad",
+          new Object[] {hostName, cpu}));
+    }
+
+    return new HostSpecBuilder(new SimpleHostAvailabilityStrategy())
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessQueryHelper.java:104]: Host availability strategy should be taken from Plugin Service.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterMonitor.java [hunk 9]
```diff
@@ -58,11 +60,12 @@ public class LimitlessRouterMonitor implements AutoCloseable, Runnable {
   private static final String MONITORING_PROPERTY_PREFIX = "limitless-router-monitor-";
   private final int intervalMs;
   private final @NonNull HostSpec hostSpec;
+  private final @NonNull String limitlessRouterEndpointQuery;
   private final AtomicBoolean stopped = new AtomicBoolean(false);
-  private final AtomicReference<List<HostSpec>> limitlessRouters = new AtomicReference<>(
-      Collections.unmodifiableList(new ArrayList<>()));
+  private final AtomicReference<List<HostSpec>> limitlessRouters;
   private final @NonNull Properties props;
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterMonitor.java:66]: protected?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java [hunk 10]
```diff
@@ -51,26 +56,64 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
...
+      final List<HostSpec> limitlessRouters =
+          limitlessRouterCache.get(this.pluginService.getHostListProvider().getClusterId(), cacheExpirationNano);
+      if (!Utils.isNullOrEmpty(limitlessRouters)) {
+        return limitlessRouters;
+      }
+
+      final List<HostSpec> newLimitLessRouters =
+          this.queryHelper.queryForLimitlessRouters(connection, hostPort);
+
+      limitlessRouterCache.remove(this.pluginService.getHostListProvider().getClusterId());
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:107]: Why do we need to remove the entry?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java [hunk 11]
```diff
@@ -51,26 +56,64 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
...
+      limitlessRouterCache.computeIfAbsent(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:108]: Can we use a simple put()?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java [hunk 12]
```diff
@@ -98,20 +141,30 @@ public List<HostSpec> forceGetLimitlessRouters(final String clusterId, final Pro
...
           key -> this.limitlessRouterMonitorInitializer
-              .createLimitlessRouterMonitor(pluginService, hostSpec, props, intervalMs),
+              .createLimitlessRouterMonitor(
+                  pluginService,
+                  hostSpec,
+                  limitlessRouters,
+                  props,
+                  intervalMs),
           cacheExpirationNano);
+      limitlessRouterCache.computeIfAbsent(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:163]: What's the purpose to add an empty list to the cache?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java (hunk 10 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java]: updated SlidingExpirationCache with `put()` method

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java (hunk 12 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java]: Was trying something dumb. 
Updated so monitors have access to the router cache directly and can update it directly.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 13]
```diff
@@ -171,36 +195,70 @@ private Connection connectInternal(
...
   private void initLimitlessRouterMonitorService() {
     if (limitlessRouterService == null) {
       this.limitlessRouterService = this.limitlessRouterServiceSupplier.get();
     }
   }
 
-  private Connection retryConnectWithLeastLoadedRouters(final List<HostSpec> limitlessRouters,
-      final Properties props, final SQLException originalException) throws SQLException {
+  private Connection retryConnectWithLeastLoadedRouters(final List<HostSpec> limitlessRouters, final Properties props,
+      final Connection conn, final HostSpec hostSpec, final SQLException originalException) throws SQLException {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:243]: Do we actually need to keep an original exception? What's the value it gives to a user to know that least-loaded-strategy has been caused by some node unavailability?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 14]
```diff
@@ -171,36 +195,70 @@ private Connection connectInternal(
...
     while (retryCount++ < maxRetries) {
-      if (!currentRouters.stream().anyMatch(h -> h.getAvailability().equals(HostAvailability.AVAILABLE))) {
-        currentRouters = synchronouslyGetLimitlessRoutersWithRetry(props);
+      if (currentRouters.stream().noneMatch(h -> h.getAvailability().equals(HostAvailability.AVAILABLE))) {
+        currentRouters = synchronouslyGetLimitlessRoutersWithRetry(conn, hostSpec.getPort(), props);
         if (currentRouters == null
             || currentRouters.isEmpty()
-            || !currentRouters.stream().anyMatch(h -> h.getAvailability().equals(HostAvailability.AVAILABLE))) {
+            || currentRouters.stream().noneMatch(h -> h.getAvailability().equals(HostAvailability.AVAILABLE))) {
           throw new SQLException(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:255]: Does it make sense to connect to initialHostSpec then?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 15]
```diff
@@ -225,16 +283,16 @@ private Connection retryConnectWithLeastLoadedRouters(final List<HostSpec> limit
     throw new SQLException(Messages.get("LimitlessConnectionPlugin.noRoutersAvailableForRetry"), originalException);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:283]: We also need to catch SQLException caused by strategy been unable to get a host and continue with our retry loop.

Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:283]: Does it make sense to connect to initialHostSpec then?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java [hunk 16]
```diff
@@ -51,55 +55,66 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
...
         MONITOR_DISPOSAL_TIME_MS.getLong(props));
-
-    final LimitlessRouterMonitor
-        limitlessRouterMonitor = limitlessRouterMonitors.get(clusterId, cacheExpirationNano);
-    if (limitlessRouterMonitor == null) {
-      throw new SQLException(
-          Messages.get("LimitlessRouterServiceImpl.nulLimitlessRouterMonitor", new Object[]{clusterId}));
+    final List<HostSpec> limitlessRouters = limitlessRouterCache.get(clusterId, cacheExpirationNano);
+    if (limitlessRouters == null) {
+      return Collections.emptyList();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:85]: return null?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java [hunk 17]
```diff
@@ -51,55 +55,66 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
...
     final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
         MONITOR_DISPOSAL_TIME_MS.getLong(props));
 
-    final LimitlessRouterMonitor
-        limitlessRouterMonitor = limitlessRouterMonitors.get(clusterId, cacheExpirationNano);
-    if (limitlessRouterMonitor == null) {
-      throw new SQLException(
-          Messages.get("LimitlessRouterServiceImpl.nulLimitlessRouterMonitor", new Object[]{clusterId}));
-    }
     forceGetLimitlessRoutersLock.lock();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:96]: That effectively locks all requests for a router list even these requests are for different clusters.

File: wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java [hunk 18]
```diff
@@ -95,6 +95,18 @@ public V computeIfAbsent(
     return cacheItem.withExtendExpiration(itemExpirationNano).item;
   }
 
+  public V put(
+      final K key,
+      final V value,
+      final long itemExpirationNano) {
+    cleanUp();
+    final CacheItem cacheItem =  cache.put(key, new CacheItem(value, itemExpirationNano));
+    if (cacheItem == null) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java:104]: That never happens.

Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java:104]: I was a bit surprised by this, but I've observed it happens on the first put (and im guessing if that k/v pair has been previously cleared).

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java (hunk 13 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:243]: Good point. Will remove it.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java (hunk 14 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:255]: Good catch. At this point, the current connection should be working.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java (hunk 17 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java:96]: That is not ideal. 
I havent fully fleshed it out, but Im currently looking at Conditions to see if they might work for this. 
https://www.baeldung.com/java-concurrent-locks#working-with-conditions

Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java]: Update: used a map of locks with cluster IDs as keys.

**Review by sergiyvamz — APPROVED** [review]

**Pull request merged**